libuavcan/src/transport/can_acceptance_filter_configurator.cpp

```diff
@@ -0,0 +1,256 @@
+                     multiset_configs_.getByIndex(i)->id);
+        UAVCAN_TRACE("CanAcceptanceFilterConfigurator::fillArray()", "map.MK [%u] = %d", i,
+                     multiset_configs_.getByIndex(i)->mask);
+    }
+
+    return 0;
+}
+
+int16_t CanAcceptanceFilterConfigurator::computeConfiguration()
+{
+    uint16_t i_rank = 0, j_rank = 0;
+    const uint16_t acceptance_filters_number = getNumFilters();
+
+    while (acceptance_filters_number < multiset_configs_.getSize())
+    {
+        UAVCAN_ASSERT(multiset_configs_.getSize() != 0);
+
+        uint8_t best_rank = 0;
+        const uint16_t map_array_size = static_cast<uint16_t>(multiset_configs_.getSize());
+
+        for (uint16_t i_ind = 0; i_ind < map_array_size - 1; i_ind++)
+        {
+            for (uint16_t j_ind = static_cast<uint16_t>(i_ind + 1); j_ind < map_array_size; j_ind++)
+            {
+                CanFilterConfig temp_array = mergeFilters(*multiset_configs_.getByIndex(i_ind),
+                                                          *multiset_configs_.getByIndex(j_ind));
+                uint8_t rank = countBits(temp_array.mask);
+                if (rank > best_rank)
+                {
+                    best_rank = rank;
+                    i_rank = i_ind;
+                    j_rank = j_ind;
+                }
+            }
+        }
+
+        *multiset_configs_.getByIndex(j_rank) = mergeFilters(*multiset_configs_.getByIndex(i_rank),
+                                                             *multiset_configs_.getByIndex(j_rank));
+        CanFilterConfig temp_key = *multiset_configs_.getByIndex(i_rank);
+
+        if ((multiset_configs_.getByIndex(i_rank)->id == multiset_configs_.getByIndex(j_rank)->id) &&
+            (multiset_configs_.getByIndex(i_rank)->mask == multiset_configs_.getByIndex(j_rank)->mask))
+        {
+            multiset_configs_.removeAll(temp_key);             // If i_rank and j_rank elements are the same, we need to add it
+            multiset_configs_.emplace(temp_key);               //another time, to preclude removal of both elements
+        }
+        else
+        {
+            multiset_configs_.removeAll(temp_key);
+        }
+    }
+
+    UAVCAN_ASSERT(acceptance_filters_number >= multiset_configs_.getSize());
+
+    return 0;
+}
+
+int16_t CanAcceptanceFilterConfigurator::applyConfiguration(void)
+{
+    CanFilterConfig filter_conf_array[MaxCanAcceptanceFilters];
+    const unsigned int filter_array_size = multiset_configs_.getSize();
+
+    if (filter_array_size > MaxCanAcceptanceFilters)
+    {
+        UAVCAN_ASSERT(0);
+        return -ErrLogic;
+    }
+
+    for (uint16_t i = 0; i < filter_array_size; i++)
+    {
+        CanFilterConfig temp_filter_config = *multiset_configs_.getByIndex(i);
+        if (multiset_configs_.getByIndex(i) == NULL)
+        {
+            UAVCAN_ASSERT(0);
+        }
+
+        if ((temp_filter_config.id != 0) || (temp_filter_config.mask != 0))
```

Comment:
> theoretically mask might be equal to zero in case there are much more listeners than available filters, so it would merge until all digits go zero.

Valid point, though a configuration that has at least one zero mask is meaningless (it won't filter anything). It's not actually a problem of the algorithm itself, it merely means that there's no solution for the provided inputs. It's just an observation, I'm not proposing any changes.